### PR TITLE
Implement MultiHeadAttention Layer

### PR DESCRIPTION
**Pull request by pforderique**
Implements the MultiHeadAttention layer from [Keras](https://github.com/keras-team/keras/blob/v2.13.1/keras/layers/attention/multi_head_attention.py) attention layers.

NOTE:
* This implementation does not support RaggedTensors yet.
* The `Softmax` layer was changed to support masking. Let me know if having a separate class for this is better. Regardless, the current change should be non-breaking.

Depends on #7860.

### CODE REVIEW

**Review by mattsoulanille — APPROVED** [review]
LGTM

**Comment by mattsoulanille** [comment]
@Linchenn Please take a look when you get a chance. Thanks!

**Review by Linchenn — APPROVED** [review]
LGTM!
